src/signTx.c

```diff
@@ -434,7 +432,95 @@ static void signTx_handleOutput_ui_runStep()
 		        this_fn
 		);
 	}
-	UI_STEP(HANDLE_OUTPUT_STEP_RESPOND) {
+	UI_STEP(HANDLE_OUTPUT_ADDRESS_STEP_RESPOND) {
+		respondSuccessEmptyMsg();
+
+		// Advance state to next output
+		ctx->currentOutput++;
+		if (ctx->currentOutput == ctx->numOutputs) {
+			advanceStage();
+		}
+	}
+	UI_STEP_END(HANDLE_OUTPUT_ADDRESS_STEP_INVALID);
+}
+
+static void signTx_handleOutput_address()
+{
+	ASSERT(ctx->stageData.output.outputType == SIGN_TX_OUTPUT_TYPE_ADDRESS);
+
+	security_policy_t policy = policyForSignTxOutputAddress(
+	                                   ctx->stageData.output.addressBuffer, ctx->stageData.output.addressSize,
+	                                   ctx->networkId, ctx->protocolMagic
+	                           );
+	TRACE("Policy: %d", (int) policy);
+	ENSURE_NOT_DENIED(policy);
+
+	{
+		// add to tx
+		ASSERT(ctx->stageData.output.addressSize > 0);
+		ASSERT(ctx->stageData.output.addressSize < BUFFER_SIZE_PARANOIA);
+
+		txHashBuilder_addOutput(
+		        &ctx->txHashBuilder,
+		        ctx->stageData.output.addressBuffer,
+		        ctx->stageData.output.addressSize,
+		        ctx->stageData.output.amount
+		);
+	}
+
+	{
+		// select UI steps
+		switch (policy) {
+#	define  CASE(POLICY, UI_STEP) case POLICY: {ctx->ui_step=UI_STEP; break;}
+			CASE(POLICY_SHOW_BEFORE_RESPONSE, HANDLE_OUTPUT_ADDRESS_STEP_DISPLAY_AMOUNT);
+			CASE(POLICY_ALLOW_WITHOUT_PROMPT, HANDLE_OUTPUT_ADDRESS_STEP_RESPOND);
+#	undef   CASE
+		default:
+			THROW(ERR_NOT_IMPLEMENTED);
+		}
+	}
+	signTx_handleOutput_address_ui_runStep();
+}
+
+enum {
+	HANDLE_OUTPUT_ADDRESSPARAMS_STEP_DISPLAY_AMOUNT = 350,
+	HANDLE_OUTPUT_ADDRESSPARAMS_STEP_DISPLAY_SPENDING_PATH,
+	HANDLE_OUTPUT_ADDRESSPARAMS_STEP_DISPLAY_STAKING_INFO,
+	HANDLE_OUTPUT_ADDRESSPARAMS_STEP_RESPOND,
+	HANDLE_OUTPUT_ADDRESSPARAMS_STEP_INVALID,
+};
+
+static void signTx_handleOutput_addressParams_ui_runStep()
+{
+	TRACE("UI step %d", ctx->ui_step);
+	ui_callback_fn_t* this_fn = signTx_handleOutput_addressParams_ui_runStep;
+
+	UI_STEP_BEGIN(ctx->ui_step);
+
+	UI_STEP(HANDLE_OUTPUT_ADDRESSPARAMS_STEP_DISPLAY_AMOUNT) {
+		ui_displayAmount(ctx->stageData.output.amount, this_fn);
+	}
+	UI_STEP(HANDLE_OUTPUT_ADDRESSPARAMS_STEP_DISPLAY_SPENDING_PATH) {
+		char pathStr[100];
+		{
```

Comment:
In the similar style -- don't we want to have ui_displaySpendingPath helper? I actually like the idea of such specialized ui utilities. Maybe we can collect it in one file uiScreens ...